### PR TITLE
[SPARK-32703][SQL] Replace deprecated API calls from SpecificParquetRecordReaderBase

### PR DESCRIPTION
**Pull request by sunchao**
### What changes were proposed in this pull request?

Currently in `SpecificParquetRecordReaderBase` we use deprecated APIs in a few places from Parquet, such as `readFooter`, `ParquetInputSplit`, `new ParquetFileReader`, `filterRowGroups`, etc. This replaces these with the newer APIs. In specific this:
- Replaces `ParquetInputSplit` with `FileSplit`. We never use specific things in the former such as `rowGroupOffsets` so the swap is pretty simple.
- Removes `readFooter` calls by using `ParquetFileReader.open`
- Replace deprecated `ParquetFileReader` ctor with the newer API which takes `ParquetReadOptions`.
- Removes the unnecessary handling of case when `rowGroupOffsets` is not null. It seems this never happens.

### Why are the changes needed?

The aforementioned APIs were deprecated and is going to be removed at some point in future. This is to ensure better supportability. 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

This is a cleanup and relies on existing tests on the relevant code paths.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you, @sunchao !

**Comment by sunchao** [comment]
Seems test is flaky - let me re-trigger it.

**Comment by maropu** [comment]
I'm running TPCDS queries based on this fix.

**Comment by dongjoon-hyun** [comment]
Thank you, @maropu .

**Comment by maropu** [comment]
okay, I've checked that there is no perf regression caused by this PR. Nice, @sunchao !

**Review by srowen — APPROVED** [review]

**Comment by maropu** [comment]
LGTM

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java [hunk 1]
```diff
@@ -227,6 +189,8 @@ protected void initialize(String path, List<String> columns) throws IOException
       }
     }
     this.sparkSchema = new ParquetToSparkSchemaConverter(config).convert(requestedSchema);
+    // unfortunately we'd have to create the reader again since there is no column projection
+    // in the new API.
     this.reader = new ParquetFileReader(
```
Comment (LuciferYang) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java:194]: Can we reuse the `ParquetFileReader ` create at line 166 and manual  set `requestedSchema ` ?

Comment (sunchao) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java:194]: You're right. I missed this. Will update.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**